### DISCUSSION TITLE
EVPN VLAN-aware bundle service data model

### DISCUSSION THREAD

**Discussion by ipspace** [ipspace/netlab#554] [Features]
I encountered an interesting challenge while refactoring EVPN code: what attributes do we want to see on VLANs in a VLAN bundle.

The refactored code assumes that every EVPN-managed object should have **evpn.evi** which is then used to generate **rd** (when not already generated). In the case of individual VLANs, **evpn.evi** is set to **vlan.id**, in the case of VLAN bundle, all VLANs are in the same MAC-VRF, so **evpn.evi** is set to the **vrf.id**. For VLANs in the bundle, the code also sets **evpn.bundle** which points to the VRF.

The existing EVPN templates assume that the presence of **evpn.evi** in a VLAN indicates it uses VLAN-based service interface. When trying to configure that service, the template fails because the VLAN does not have **evpn.import** or **evpn.export** attributes (because it's part of a bundle).

I would prefer to fix the configuration templates (EOS and SR Linux) to check **evpn.bundle**, and keep **evpn.evi** in the VLAN data. Any other preferences @jbemmel?

**Comment by jbemmel** [comment]
No problem with that

**Comment by jbemmel** [comment]
Turns out I do have another preference, as SR OS does not support the VLAN-aware bundle service - only the [VLAN bundle service](https://datatracker.ietf.org/doc/html/rfc7432#section-6.2)

Can we add a feature flag for vlan-bundle and vlan-aware-bundle services, and support both models in the data model?

**Comment by ipspace** [comment]
Well, if I was stupid enough to implement VLAN-aware bundle service that's supported on two platforms, I don't have the high moral ground to say NO because VLAN bundle service is supported only on one platform ;)... but who knows, we might add Junos and IOS XR in the future ;))

Explore how much the data model would have to change and we'll go from there.

**Comment by jbemmel** [comment]
I think we should model what each device supports ( e.g. EOS vlan-aware-bundle, SROS vlan-bundle, JUNOS both ).
Then we could auto-select which service model to implement based on the selected devices, and allow the user to specify a specific service model and throw an error if the user tries to do vlan-aware-bundle with a device that doesn't support it (or vice versa)

Proposal:
```
evpn.bundle:
  vrf: xyz
  vlan_aware: True # Omit to auto-select based on device supported feature(s)
```

```
features:
 evpn:
  vlan_bundle: True
  vlan_aware_bundle: False
```
